r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+#'    decay (value of 1). Propagates as mean**(post_treat_period).
+#' @param decay_sd_inp Standard deviation of the truncated normal decay.
+#'
+#' @return Tibble with all individual level work completed.
+.UnitLevelSimulation <- function(n_inp,
+                                 type_inp,
+                                 cov_overlap_inp,
+                                 loading_scale_inp,
+                                 num_factors_inp,
+                                 int_scale_inp,
+                                 rho_inp,
+                                 rho_scale_inp,
+                                 rho_shift_inp,
+                                 prop_treated_inp,
+                                 treat_start,
+                                 num_periods_inp,
+                                 impact_mean_inp,
+                                 impact_sd_inp,
+                                 decay_mean_inp,
+                                 decay_sd_inp) {
+  # Gather data that is time invariant.
+  # First, call to assign treatment, which generates covariates.
+  unit_level_tib <- .AssignTreat(
+    n_inp = n_inp,
+    type_inp = type_inp,
+    cov_overlap_inp = cov_overlap_inp,
+    loading_scale_inp = loading_scale_inp,
+    num_factors_inp = num_factors_inp,
+    rho_inp = rho_inp,
+    rho_scale_inp = rho_scale_inp,
+    rho_shift_inp = rho_shift_inp,
+    int_scale_inp = int_scale_inp,
+    prop_treated_inp = prop_treated_inp
+  )
+
+  # Merge in the treatment period assignment.
+  unit_level_tib <- .AssignTreatTime(
+    unit_level_tib, treat_start,
+    num_periods_inp
+  )
+  # Allow wiggle room for the TE without getting NA.
+  impact_ub <- ifelse(impact_mean_inp < 0.25, 0.25,
+    min(1, impact_mean_inp + (0.25))
+  )
+  impact_lb <- ifelse(impact_mean_inp <= 0, impact_mean_inp - 0.25, 0)
+  decay_ub <- ifelse(decay_mean_inp > 0.9, 1.000000001, 0.9)
+  # Assign treatment effect impact and decay parameters per unit.
+  unit_level_tib <- unit_level_tib %>%
+    dplyr::mutate(
+      treat_impact = truncnorm::rtruncnorm(
+        n = dplyr::n(), a = impact_lb, b = impact_ub,
+        mean = impact_mean_inp, sd = impact_sd_inp
+      ),
+      treat_decay = truncnorm::rtruncnorm(
+        n = dplyr::n(), a = 0, b = decay_ub,
+        mean = decay_mean_inp, sd = decay_sd_inp
+      )
+    )
+  # Require that no treat decay is larger than 1.
+  unit_level_tib <- unit_level_tib %>%
+    dplyr::mutate(treat_decay = dplyr::case_when(
+      treat_decay > 1 ~ 1,
+      treat_decay <= 1 ~ treat_decay
+    ))
+
+  return(unit_level_tib)
+}
+
+#' Treatment assignment mechanism, with shifted covariates appended.
+#'
+#' @inheritParams .UnitLevelSimulation
+#'
+#' @return Tibble with treatment assignment and individual specific covariates
+#'    shifted conditional on treatment assignment and inputs.
+.AssignTreat <- function(n_inp, type_inp, cov_overlap_inp,
```

Comment:
`type_inp`'s value should be checked using `match.arg`.

Reply:
Repeat of prior comment: Is that redundant? The user inputs type, which we match.arg at the top level. Then every helper function gets passed that match.arg'ed (type_inp) parameter. I don't have a strong sense of the proper way to do this, so I will follow your judgement!